File: codeforces/000/712D.py
#
# 712D
# TLE
#

M = 1000000007

a, b, k, t = None, None, None, None
dp = None
offset = None
idx_max = None
delta_min, delta_max = None, None

def idx(x):
	return x + offset

def dpfill(i, j):
	if j == idx_max - k:
		dp[i][j] = reduce(lambda x, y: x + y, dp[i+1][j-k : j+k+1]) % M
	else:
		dp[i][j] = (dp[i+1][j-k] + dp[i][j+1] - dp[i+1][j+k+1]) % M
	#print "fill: dp[%2d][%2d]: %d"%(i, j - offset, dp[i][j])

if __name__ == "__main__":
	a, b, k, t = map(lambda x: int(x), raw_input().split())
	#a, b, k, t = 1, 2, 2, 1
	#a, b, k, t = 1, 1, 1, 2
	#a, b, k, t = 2, 7, 3, 1

	t *= 2
	delta = a - b
	delta_max = delta + t * k
	delta_min = delta - t * k
	offset = -1 * delta_min
	idx_max = delta_max + offset

	#print "offset: ", offset

	dp = [[0 for j in range(idx_max + 1)] for i in range(t + 1)]

	for j in range(0, idx_max + 1):
		dp[t][j] = k - offset + j
		if dp[t][j] < 0:
			dp[t][j] = 0
		if dp[t][j] > 2 * k + 1:
			dp[t][j] = 2 * k + 1
		dp[t][j] %= M
		#print "dp[%2d][%2d]: %d"%(t, j - offset, dp[t][j])

	for i in range(t - 1, 0, -1):
		for j in range(delta_max - k, delta_min + k - 1, -1):
			dpfill(i, idx(j))

	print dp[1][delta + offset]

File: codeforces/000/712D.c
/**
 * 712D
 */
#include "common.h"
#include "numth.h"

#define M 1000000007
#define T 100
#define K 1000

#define pls(a, b) NUMTH_MODPLS((a), (b), M)

int dp[2*T][4*T*K+1];
int a, b, t, k;
int delta, delta_min, delta_max, offset, index_max;

int dp_fill(int i, int j)
{
	int x;

	if (j == index_max-k) {
		dp[i][j] = 0;
		forre(x, j-k, j+k) dp[i][j] = pls(dp[i][j], dp[i+1][x]);
	} else {
		dp[i][j] = pls(pls(dp[i+1][j-k], dp[i][j+1]), -1*dp[i+1][j+k+1]);
		NUMTH_CORRECT(dp[i][j], M);
	}

	//print("[i][i]: i", i, j, dp[i][j]);
	return dp[i][j];
}

int main(void)
{
	int i, j; 

	//scanf("%d%d%d%d", &a, &b, &t, &k);
	a = 1, b = 1, k = 1, t = 2; //31
	//a = 1, b = 2, k = 2, t = 1; //6
	//a = 2, b = 12, k = 3, t = 1;

	t <<= 1;
	delta = a - b;
	delta_min = delta - t * k;
	delta_max = delta + t * k;
	offset = -1 * delta_min;
	index_max = delta_max + offset;

	forre(i, 0, index_max) dp[t-1][i] = max(0, k-max(-1*k, offset-i+1)+1);
	//forre(i, 0, index_max) print("[i][i]: i", t-1, i, dp[t-1][i]);

	/* FIXME:
	 * There are TWO positions that will use delta_max + offset!!!
	 *
	 * i = t-2, j = delta_max-k+offset;
	 * dp[i][j] = 0;
	 * forre(x, index_max-2*k, index_max) dp[i][j] = pls(dp[i][j], dp[i+1][x]);
	 * print("[i][i]: i", i, j, dp[i][j]);
	 */

	for (i = t-2; i >= 0; --i)
		for (j = index_max-k; j >= k; --j)
			dp_fill(i, j);

	printf("%d\n", dp[0][delta+offset]);
	return 0;
}
